metadata-ingestion/src/datahub/specific/dataset.py

```diff
@@ -144,6 +147,89 @@ def set_upstream_lineages(self, upstreams: List[Upstream]) -> "DatasetPatchBuild
         )
         return self
 
+    def add_fine_grained_upstream_lineage(
+        self, fine_grained_lineage: FineGrainedLineage
+    ) -> "DatasetPatchBuilder":
+        transform_op = (
+            fine_grained_lineage.transformOperation
+            if fine_grained_lineage.transformOperation is not None
+            else "NONE"
+        )
+        upstream_type = (
+            fine_grained_lineage.upstreamType
+            if isinstance(fine_grained_lineage.upstreamType, str)
+            else FineGrainedLineageUpstreamType.FIELD_SET
+        )
+        downstream_type = (
+            fine_grained_lineage.downstreamType
+            if isinstance(fine_grained_lineage.downstreamType, str)
+            else FineGrainedLineageDownstreamType.FIELD_SET
+        )
+        for upstream_urn in fine_grained_lineage.upstreams or []:
+            self._add_patch(
+                UpstreamLineage.ASPECT_NAME,
+                "add",
+                path=f"/fineGrainedLineages/{quote(transform_op, safe='')}/upstreamType/"
+                f"{quote(upstream_type, safe='')}/{quote(upstream_urn, safe='')}",
+                value=fine_grained_lineage.confidenceScore,
+            )
+        for downstream_urn in fine_grained_lineage.downstreams or []:
+            self._add_patch(
+                UpstreamLineage.ASPECT_NAME,
+                "add",
+                path=f"/fineGrainedLineages/{quote(transform_op, safe='')}/downstreamType/"
+                f"{quote(downstream_type, safe='')}/{quote(downstream_urn, safe='')}",
+                value=fine_grained_lineage.confidenceScore,
+            )
+        return self
+
+    def remove_fine_grained_upstream_lineage(
+        self, fine_grained_lineage: FineGrainedLineage
+    ) -> "DatasetPatchBuilder":
+        transform_op = (
+            fine_grained_lineage.transformOperation
+            if fine_grained_lineage.transformOperation is not None
+            else "NONE"
+        )
+        upstream_type = (
+            fine_grained_lineage.upstreamType
+            if isinstance(fine_grained_lineage.upstreamType, str)
+            else FineGrainedLineageUpstreamType.FIELD_SET
+        )
+        downstream_type = (
+            fine_grained_lineage.downstreamType
+            if isinstance(fine_grained_lineage.downstreamType, str)
+            else FineGrainedLineageDownstreamType.FIELD_SET
+        )
```

Comment:
some of this code looks copy-pasted - can we refactor into a static method

ideally it'd be `transform_op, upstream_type, downstream_type = self._preprocess_cll_patch(fine_grained_lineage)` or something similar